### PR TITLE
added solution

### PR DESCRIPTION
**Pull request by dev-dwr**

### CODE REVIEW

File: config/default.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+export default {
+  port: 4000,
+  dbUri: "mongodb://localhost:27017/mongodb",
+  logLevel: "info",
+  smtp: {
+    user: "xrynelkxp7raewvg@ethereal.email",
+    pass: "QSXe1fFNU7qZHwR4Xc",
```
Comment (Mavannkas) [inline config/default.ts:7]: Do not push credits to repo!!!!! NEVER

File: data/insertSeats.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  try {
+    await TablesModel.insertMany(seats);
+    log.info("Tables have been loaded");
+  } catch (e: any) {
+    log.error(e, "could not load seats");
+  }
+}
+
+export default loadSeatsData;
```
Comment (Mavannkas) [inline data/insertSeats.ts:20]: Better than default export is a named export. The code editor will help you more precisely, you only use what you currently need. And you can export a lot of things. (You import them with destructuring)

File: data/insertSeats.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import fs from "fs";
+import TablesModel from "../src/models/tables.model";
+import log from "../src/utils/logger";
+import path from "path";
+
+async function loadSeatsData() {
+  const seatsData = fs.readFileSync(path.resolve(__dirname + "/seats.json"), {
```
Comment (Mavannkas) [inline data/insertSeats.ts:7]: Name of the file should in CONST variable and you should use path join

File: src/controllers/reservation.controller.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  const table = await findTableBySeatNumber(seatNumber);
+
+  if (reservationExists)
+    return res
+      .status(400)
+      .json({ error: "This table is alredy taken in this time" });
+  if (!table)
+    return res.status(404).send("Table of given number does not exist");
+
+  if (numberOfSeats > table.maxNumberOfSeats)
```
Comment (Mavannkas) [inline src/controllers/reservation.controller.ts:42]: You should always use full if structure, especially when the body is on several lines

File: src/controllers/reservation.controller.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  deleteReservationById
+} from "../services/reservation.service";
+import { findTableBySeatNumber } from "../services/table.service";
+import sendEmail from "../utils/mailer";
+import log from "../utils/logger";
+import { customAlphabet } from "nanoid";
+export async function createReservationHandler(
+  req: Request<{}, {}, CreateReservationInput>,
+  res: Response
+) {
```
Comment (Mavannkas) [inline src/controllers/reservation.controller.ts:24]: Too long function, don't forget the single responsibility principle

File: src/controllers/reservation.controller.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+  if (numberOfSeats > table.maxNumberOfSeats)
+    return res
+      .status(400)
+      .json({ error: "It is too many seats for this table" });
+
+  if (numberOfSeats < table.minNumberOfSeats)
+    return res.status(400).json({ error: "The amount of seats is too small" });
+
+  try {
```
Comment (Mavannkas) [inline src/controllers/reservation.controller.ts:50]: When you use try catch the best practice is to move the body to a separate function.

File: src/controllers/reservation.controller.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    const createdReservation = await createReservation(req.body);
+
+    table.status = "TAKEN";
+    await table.save();
+
+    await sendEmail({
+      from: "test@example.com",
+      to: email,
+      subject: "Your reservation details",
+      text: `Hello ${fullName} Reservation Id: ${
```
Comment (Mavannkas) [inline src/controllers/reservation.controller.ts:60]: Better to use a separate function to generate text or email body

File: src/controllers/reservation.controller.ts [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        createdReservation._id
+      }, Date of reservation: ${moment(date).format(
+        "dddd, MMMM Do YYYY, h:mm:ss a"
+      )}, Table: ${seatNumber}, 
+      number of seats: ${numberOfSeats}`,
+    });
+    return res.status(201).json({ reservationId: createdReservation._id });
+  } catch (e: any) {
+    log.error(e, "Error occured while creating reservation");
+    return res.status(500).send(e);
```
Comment (Mavannkas) [inline src/controllers/reservation.controller.ts:70]: Don't send the entire error, just its msg

File: src/controllers/reservation.controller.ts [hunk 9]
```diff
@@ -0,0 +1,159 @@
+import moment from "moment";
```
Comment (Mavannkas) [inline src/controllers/reservation.controller.ts:1]: Moment is legacy lib https://momentjs.com/docs/#/-project-status/

File: src/controllers/reservation.controller.ts [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+    if (reqeustCancellationTime > twoHourseBeforeReservationTime) {
+      return res
+        .status(409)
+        .json({ error: "You cannot cancell reservation now it is to late" });
+    }
+
+    reservation.reservationStatus = cancellationStatus;
+    reservation.verificationCode = cacancellationId;
+    reservation.save();
```
Comment (Mavannkas) [inline src/controllers/reservation.controller.ts:114]: Did you forget about await?

File: src/controllers/reservation.controller.ts [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    log.error(e, "cannot send reservation cancellation request");
+    return res.status(400).send(e.message);
+  }
+}
+
+export async function deleteReservationHandler(
+  req: Request<DeleteRequestParams, {}, DeleteRequestInput>,
+  res: Response
+) {
+  const { id } = req.params;
```
Comment (Mavannkas) [inline src/controllers/reservation.controller.ts:133]: Line 90, try to follow one pattern

File: src/controllers/table.controller.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    const reservationsOfGivenTime = await findReservations({ date, duration });
+    const takenSeatsInGivenTime: number[] = [];
+    reservationsOfGivenTime.map((reservation) =>
+      takenSeatsInGivenTime.push(reservation.seatNumber)
+    );
+    const result: getFreeTablesResult[] = [];
+
+    tables.forEach((table) => {
+      if (!takenSeatsInGivenTime.includes(table.number))
+        return result.push(omit(table.toJSON(), ["_id", "__v", "status"]));
```
Comment (Mavannkas) [inline src/controllers/table.controller.ts:34]: Better to use map Instead of a slow lodash

File: src/controllers/table.controller.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+interface getFreeTablesResult {
+  number: number;
+  minNumberOfSeats: number;
+  maxNumberOfSeats: number;
+}
+export async function getFreeTablesHandler(req: Request, res: Response) {
+  const qsDate = req.query.start_date as string;
+  const date = new Date(qsDate);
+  const durationQs = req.query.duration as string;
+  const duration = parseInt(durationQs);
```
Comment (Mavannkas) [inline src/controllers/table.controller.ts:15]: `parseint("1234") = +"1234"`

File: src/controllers/table.controller.ts [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+ 
+  try {
+    const allTablesOfStatusFree = await findTables(status);
+    const tables = allTablesOfStatusFree.filter(
+      (table) => table.minNumberOfSeats >= parseInt(min_seats)
+    );
+
+    const reservationsOfGivenTime = await findReservations({ date, duration });
+    const takenSeatsInGivenTime: number[] = [];
+    reservationsOfGivenTime.map((reservation) =>
```
Comment (Mavannkas) [inline src/controllers/table.controller.ts:27]: Does the map have any sense in this use case? Better use forEach

File: src/controllers/reservation.controller.ts [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+      .json({ error: "It is too many seats for this table" });
+
+  if (numberOfSeats < table.minNumberOfSeats)
+    return res.status(400).json({ error: "The amount of seats is too small" });
+
+  try {
+    const createdReservation = await createReservation(req.body);
+
+    table.status = "TAKEN";
+    await table.save();
```
Comment (Mavannkas) [inline src/controllers/reservation.controller.ts:54]: Logic should be in service

File: src/utils/mailer.ts [hunk 16]
```diff
@@ -0,0 +1,31 @@
+import nodemailer, { SendMailOptions } from "nodemailer";
```
Comment (Mavannkas) [inline src/utils/mailer.ts:1]: I think email body should be here

**Comment by dev-dwr** [comment]
thanks for the review

**Pull request closed without merging**